table/multiget_context.h

```diff
@@ -0,0 +1,238 @@
+// easier by allowing various stages of the MultiGet lookups to operate on
+// subsets of keys, potentially non-contiguous. In order to accomplish this,
+// it defines the following classes -
+//
+// MultiGetContext::Range - Specifies a range of keys, by start and end index,
+// from the parent MultiGetContext. Each range contains a bit vector that
+// indicates whether the corresponding keys need to be processed or skipped.
+// A Range object can be copy constructed, and the new object inherits the
+// original Range's bit vector. This is useful for progressively skipping
+// keys as the lookup goes through various stages. For example, when looking
+// up keys in the same SST file, a Range is created excluding keys not
+// belonging to that file. A new Range is then copy constructed and individual
+// keys are skipped based on bloom filter lookup.
+//
+// MultiGetContext::Range::Iterator - A forward iterator that iterates over
+// non-skippable keys in a Range, as well as keys whose final value has been
+// found. The latter is tracked by MultiGetContext::value_mask_
+// MultiGetContext::Range::IteratorWrapper - A wrapper around a vector
+// container, such as std::vector, std::array or autovector, that shadows a
+//
+// MultiGetContext::Range::Iterator. The size of the vector must be atleast
+// the number of keys in the MultiGetContext batch (limited by
+// MultiGetContext::MAX_KEYS_ON_STACK), with the indexes of keys being
+// identical to that in MultiGetContext. This is useful for maintaining
+// auxillary data structures that can be quickly accessed while iterating
+// through a MultigetContext::Range. The auxillary data structures can be
+// allocated on stack for efficiency
+class MultiGetContext {
+ public:
+  // Limit the number of keys in a batch to this number. Benchmarks show that
+  // there is negligible benefit for batches exceeding this. Keeping this < 64
+  // simplifies iteration, as well as reduces the amount of stack allocations
+  // htat need to be performed
+  static const int MAX_KEYS_ON_STACK = 32;
+
+  MultiGetContext(KeyContext** sorted_keys, size_t num_keys,
+                  SequenceNumber snapshot)
+      : sorted_keys_(sorted_keys),
+        num_keys_(num_keys),
+        value_mask_(0),
+        lookup_key_ptr_(reinterpret_cast<LookupKey*>(lookup_key_buf)) {
+    int index = 0;
+
+    for (size_t iter = 0; iter != num_keys_; ++iter) {
+      sorted_keys_[iter]->lkey = new (&lookup_key_ptr_[index])
+          LookupKey(*sorted_keys_[iter]->key, snapshot);
+      sorted_keys_[iter]->ukey = sorted_keys_[iter]->lkey->user_key();
+      sorted_keys_[iter]->ikey = sorted_keys_[iter]->lkey->internal_key();
+      index++;
+    }
+  }
+
+  ~MultiGetContext() {
+    if (reinterpret_cast<char*>(lookup_key_ptr_) != lookup_key_buf) {
+      delete[] lookup_key_ptr_;
+    }
+  }
+
+ private:
+  char lookup_key_buf[sizeof(LookupKey) * MAX_KEYS_ON_STACK];
```

Comment:
Is it very large? 200+  bytes, and 32 of them?

Do we consider to reduce MAX_KEYS_ON_STACK from 30?
I also argue that 200 bytes for LookupKey is probably too large. Something like 64 may be sufficient.

Reply:
Lowering the number of LookupKeys to 16 seems like a good number based on benchmarks.